File: pandemic/model/City.py
from __future__ import annotations

from typing import List

from pandemic.model.Color import Color


class City:
    cities: List[City] = []

    def __init__(self, name: str, color: Color):
        self.name: str = name
        self.color: Color = color
        self.has_research_center = False
        self.neighbors: List[City] = []
        self.diseases: dict[Color, int] = {
            Color.YELLOW: 0,
            Color.BLACK: 0,
            Color.BLUE: 0,
            Color.RED: 0,
        }
        self.__class__.cities.append(self)

    def add_neighbors(self, *neighbors: City):
        [self.neighbors.append(c) for c in neighbors]

    def construct_research_center(self):
        self.has_research_center = True

    def __str__(self):
        return f'{self.name.ljust(12)}{self.diseases}{" (DATA_CENTER)" if self.has_research_center else ""}'

    def __repr__(self):
        return self.__str__()

File: pandemic-front/src/city/City.js
import './City.css';
import {Component} from "react";
import {Player} from "../player/Player";



export class City extends Component {
    COLOR_MAPPING = {
        "YELLOW": "yellow-disease",
        "BLACK": "black-disease",
        "BLUE": "blue-disease",
        "RED": "red-disease",
    }
    diseasesArray = []
    styles = {
        position: 'absolute',
        left: this.props.city.position.x,
        top: this.props.city.position.y
    };

    constructor(props) {
        super(props);
        console.log('cities are rendering...')

        Object.keys(this.props.city.diseases).forEach((color) => {
            for(let i = 0; i < this.props.city.diseases[color]; i++) {
                this.diseasesArray.push(color)
            }
        })
    }

    render() {
        return (
            <div style={this.styles} className='city' onClick={() => this.sendAction(this.props.city)}>
                <span className='city-name'>{this.props.city.name}</span>
                <ul className='circle-container'>
                    {
                        this.diseasesArray.map((color, index) =>
                            <li className={`disease ${this.COLOR_MAPPING[color]}`}
                                style={{'--angle': `${360/this.diseasesArray.length*index}deg`}}
                                key={index}
                            ></li>
                        )
                    }
                </ul>
                <ul className='players-container'>
                    {this.props.players.map(player => <li className='player-container'><Player key={player.id} player={player}></Player></li>)}
                </ul>
            </div>
        )
    }

    sendAction(city) {
        console.log(city)
        return null
    }


}



